### ISSUE TITLE
Dead code elimination fails when re-assigning to variable

### ISSUE THREAD

**Issue by tindzk** [scala-native/scala-native#710]
The following code does not compile:

```scala
object Main {
  import scala.scalanative.native._

  def f(): Unit = {}
  def g(f: CFunctionPtr0[Unit]): Ptr[Byte] = ???

  def main(args: Array[String]): Unit = {
    var ptr: Ptr[Byte] = null
    ptr = g(CFunctionPtr.fromFunction0(f))
  }
}
```

It works as expected when I change the assignment to:

```scala
val ptr: Ptr[Byte] = g(CFunctionPtr.fromFunction0(f))
```

The separate assignment is needed to express the following pattern in Scala Native:

```c
static void sig_handler(int fd, short event, void *arg) {
  struct event *event = *(struct event **) arg;
}

void main() {
  struct event *sigev = NULL;
  sigev = event_new(base, -1, SIGINT, sig_handler, &sigev);
}
```

(Source: http://archives.seul.org/libevent/users/Sep-2012/msg00041.html)

The compiler reports the following stack trace:

```
java.util.NoSuchElementException: key not found: Local(src,6)
        at scala.collection.MapLike$class.default(MapLike.scala:228)
        at scala.collection.AbstractMap.default(Map.scala:58)
        at scala.collection.mutable.HashMap.apply(HashMap.scala:64)
        at scala.scalanative.optimizer.analysis.UseDef$$anonfun$scala$scalanative$optimizer$analysis$UseDef$$deps$1$1.apply(UseDef.scala:90)
        at scala.scalanative.optimizer.analysis.UseDef$$anonfun$scala$scalanative$optimizer$analysis$UseDef$$deps$1$1.apply(UseDef.scala:89)
        at scala.collection.mutable.UnrolledBuffer$Unrolled.foreach(UnrolledBuffer.scala:224)
        at scala.collection.mutable.UnrolledBuffer.foreach(UnrolledBuffer.scala:130)
        at scala.scalanative.optimizer.analysis.UseDef$.scala$scalanative$optimizer$analysis$UseDef$$deps$1(UseDef.scala:89)
        at scala.scalanative.optimizer.analysis.UseDef$$anonfun$apply$4$$anonfun$apply$5.apply(UseDef.scala:114)
        at scala.scalanative.optimizer.analysis.UseDef$$anonfun$apply$4$$anonfun$apply$5.apply(UseDef.scala:112)
        at scala.collection.mutable.UnrolledBuffer$Unrolled.foreach(UnrolledBuffer.scala:224)
        at scala.collection.mutable.UnrolledBuffer.foreach(UnrolledBuffer.scala:130)
        at scala.scalanative.optimizer.analysis.UseDef$$anonfun$apply$4.apply(UseDef.scala:112)
        at scala.scalanative.optimizer.analysis.UseDef$$anonfun$apply$4.apply(UseDef.scala:111)
        at scala.collection.Iterator$class.foreach(Iterator.scala:727)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.scalanative.optimizer.analysis.UseDef$.apply(UseDef.scala:111)
        at scala.scalanative.optimizer.pass.DeadCodeElimination.onInsts(DeadCodeElimination.scala:17)
        at scala.scalanative.optimizer.Pass$class.onDefn(Pass.scala:29)
        at scala.scalanative.optimizer.pass.DeadCodeElimination.onDefn(DeadCodeElimination.scala:12)
        at scala.scalanative.optimizer.Pass$$anonfun$onDefns$1.apply(Pass.scala:19)
        at scala.scalanative.optimizer.Pass$$anonfun$onDefns$1.apply(Pass.scala:19)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
        at scala.collection.AbstractTraversable.map(Traversable.scala:105)
        at scala.scalanative.optimizer.Pass$class.onDefns(Pass.scala:19)
        at scala.scalanative.optimizer.pass.DeadCodeElimination.onDefns(DeadCodeElimination.scala:12)
        at scala.scalanative.optimizer.Optimizer$.scala$scalanative$optimizer$Optimizer$$loop$1(Optimizer.scala:64)
        at scala.scalanative.optimizer.Optimizer$$anonfun$apply$1.apply(Optimizer.scala:74)
        at scala.scalanative.optimizer.Optimizer$$anonfun$apply$1.apply(Optimizer.scala:70)
        at scala.collection.parallel.AugmentedIterableIterator$class.map2combiner(RemainsIterator.scala:120)
        at scala.collection.parallel.immutable.ParHashMap$ParHashMapIterator.map2combiner(ParHashMap.scala:79)
        at scala.collection.parallel.ParIterableLike$Map.leaf(ParIterableLike.scala:1057)
        at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
        at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
        at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
        at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
        at scala.collection.parallel.ParIterableLike$Map.tryLeaf(ParIterableLike.scala:1054)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:162)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
        at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by densh** [comment]
Thanks for the bug report!

**Comment by densh** [comment]
Fixed in #1426
